Objective-C/CBLCoreMLPredictiveModel.m

```diff
@@ -0,0 +1,538 @@
+        }
+        
+        BOOL isSupported = (desc.predictedFeatureName || outputSpecs.count == 1);
+        if (isSupported) {
+            _vnFeatureName = inputSpecs.allKeys[0];
+        } else {
+            CBLWarn(Query, @"Vision models that returns multiple VNCoreMLFeatureValueObservation "
+                            "or VNPixelBufferObservation are not supported. The regular MLModel will "
+                            "be used instead.");
+            _vnModel = nil;
+        }
+    }
+}
+
+
+#pragma mark - CBLPredictiveModel
+
+
+- (nullable CBLDictionary*) predict: (CBLDictionary*)input {
+    if (_vnModel)
+        return [self predictUsingVNModel: input];
+    else
+        return [self predictUsingMLModel: input];
+}
+
+
+- (nullable CBLDictionary*) predictUsingMLModel: (CBLDictionary*)input {
+    id spec = _model.modelDescription.inputDescriptionsByName;
+    id<MLFeatureProvider> features = [self.class featuresFromDictionary: input spec: spec];
+    if (!features)
+        return nil;
+    
+    NSError* error;
+    id<MLFeatureProvider> output = [_model predictionFromFeatures: features error: &error];
+    if (!output)
+        return nil;
+    
+    return [self.class dictionaryFromFeatures: output];
+}
+
+
+- (nullable CBLDictionary*) predictUsingVNModel: (CBLDictionary*)input {
+    assert(_vnModel != nil && _vnFeatureName != nil);
+    
+    CBLBlob* blob = [input valueForKey: _vnFeatureName];
+    if (!blob) {
+        CBLWarn(Query, @"Cannot find blob input for vision feature named %@", _vnFeatureName);
+        return nil;
+    }
+    
+    // Process the model:
+    NSError* error;
+    VNCoreMLRequest* request = [[VNCoreMLRequest alloc] initWithModel: _vnModel];
+    VNImageRequestHandler* handler =
+        [[VNImageRequestHandler alloc] initWithData: blob.content options: @{}];
+    if (![handler performRequests: @[request] error: &error]) {
+        CBLWarn(Query, @"Failed to process vision Core ML request for vision feature named %@ : %@",
+                _vnFeatureName, error);
+        return nil;
+    }
+    
+    return [self predictionResultFromObservations: request.results];
+}
+
+
+- (CBLDictionary*) predictionResultFromObservations: (NSArray*)observations {
+    if (observations.count == 0)
+        return nil;
+    
+    // See https://developer.apple.com/documentation/vision/vncoremlrequest for
+    // the information about the VNObservation results:
+    NSString* predictedFeatureName = _model.modelDescription.predictedFeatureName;
+    if (predictedFeatureName) {
+        CBLMutableDictionary* result = [[CBLMutableDictionary alloc] init];
+        
+        // Predicted Feature:
+        VNClassificationObservation* top = observations[0];
+        [result setValue: top.identifier forKey: predictedFeatureName];
+        
+        // Predicted Probabilities:
+        NSString* predictedProbsName = _model.modelDescription.predictedProbabilitiesName;
+        if (predictedProbsName) {
```

Comment:
same
```suggestion
        if ([predictedProbsName length] > 0) {
```

Reply:
Not sure if it's necessary based on https://developer.apple.com/documentation/vision/vncoremlrequest?language=objc documentation. If that happens, it seems to an invalid model.

Reply:
So let check nil based on Apple documentation.